library/std/src/sys/fs/windows.rs

```diff
@@ -1233,74 +1233,147 @@ pub fn unlink(path: &WCStr) -> io::Result<()> {
+        }
 
-                (&raw mut (*file_rename_info).RootDirectory).write(ptr::null_mut());
-                // Don't include the NULL in the size
-                (&raw mut (*file_rename_info).FileNameLength).write(new_len_without_nul_in_bytes);
+        (&raw mut (*file_rename_info).Anonymous).write(c::FILE_RENAME_INFO_0 {
+            Flags: c::FILE_RENAME_FLAG_REPLACE_IF_EXISTS | c::FILE_RENAME_FLAG_POSIX_SEMANTICS,
+        });
 
-                new.as_ptr().copy_to_nonoverlapping(
-                    (&raw mut (*file_rename_info).FileName).cast::<u16>(),
-                    new.count_bytes(),
-                );
+        (&raw mut (*file_rename_info).RootDirectory).write(ptr::null_mut());
+        // Don't include the NULL in the size
+        (&raw mut (*file_rename_info).FileNameLength).write(new_len_without_nul_in_bytes);
+
+        new.as_ptr().copy_to_nonoverlapping(
+            (&raw mut (*file_rename_info).FileName).cast::<u16>(),
+            new.count_bytes(),
+        );
+    }
+
+    let result = cvt(unsafe {
+        c::SetFileInformationByHandle(
+            f.as_raw_handle(),
+            c::FileRenameInfoEx,
+            file_rename_info.cast::<c_void>(),
+            struct_size,
+        )
+    });
+
+    let result = match result {
+        Ok(i) => Ok(i),
+        // Windows version older than Windows 10 1607 don't support FileRenameInfoEx and fail with ERROR_INVALID_PARAMETER.
+        // On Windows 10 1607, it may fail with STATUS_VOLUME_NOT_UPGRADED, which gets mapped to ERROR_INVALID_FUNCTION.
+        // On Windows Server 2022, it may fail on ReFS with ERROR_NOT_SUPPORTED.
+        Err(err)
+            if err.raw_os_error() == Some(c::ERROR_INVALID_PARAMETER as _)
+                || err.raw_os_error() == Some(c::ERROR_INVALID_FUNCTION as _)
+                || err.raw_os_error() == Some(c::ERROR_NOT_SUPPORTED as _) =>
+        {
+            unsafe {
+                (&raw mut (*file_rename_info).Anonymous)
+                    .write(c::FILE_RENAME_INFO_0 { ReplaceIfExists: true });
             }
 
-            let result = unsafe {
+            cvt(unsafe {
                 c::SetFileInformationByHandle(
                     f.as_raw_handle(),
-                    c::FileRenameInfoEx,
+                    c::FileRenameInfo,
                     file_rename_info.cast::<c_void>(),
                     struct_size,
                 )
-            };
-            unsafe { dealloc(file_rename_info.cast::<u8>(), layout) };
-            if result == 0 {
-                if api::get_last_error() == WinError::DIR_NOT_EMPTY {
-                    return Err(WinError::DIR_NOT_EMPTY).io_result();
-                } else {
-                    return Err(err).io_result();
-                }
-            }
-        } else {
-            return Err(err).io_result();
+            })
         }
+        Err(err) => Err(err),
+    };
+
+    unsafe {
+        dealloc(file_rename_info.cast::<u8>(), layout);
```

Comment:
This is now fairly distant from the alloc. While nothing returns or panics before this atm, I wouldn't be surprised if some refactoring in the future missed this subtlety. So I would be happier with a drop guard or some other custom struct that is guaranteed to free memory wherever the scope ends.